### PR TITLE
Securely erase memory & reduce public API

### PR DESCRIPTION
**Pull request by nbougalis**

### CODE REVIEW

**Comment by HowardHinnant** [comment]
Here's one more spot that needs treatment: https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/Serializer.h#L201

File: src/ripple/protocol/impl/SecretKey.cpp [hunk 1]
```diff
@@ -199,6 +210,8 @@ generateSecretKey (KeyType type, Seed const& seed)
 
     if (type == KeyType::secp256k1)
     {
+        // FIXME: Avoid copying the seed into a uint128 key only to have
+        //        generateRootDeterministicPrivateKey copy out of it.
```
Comment (mellery451) [inline src/ripple/protocol/impl/SecretKey.cpp]: I probably don't understand all the nuances of the data - why don't we need to `secure_erase` the `upk` below?

Comment (nbougalis) [reply src/ripple/protocol/impl/SecretKey.cpp]: We'd need to massage things a bit to do it, but it's possible. Perhaps in the interest of thoroughness we should, even if it's a bit of extra work.

As I've marked elsewhere in the file, ideally we should change `generateRootDeterministicPrivateKey` to just take a `Slice` and not a `uint128`, and have it return a `SecretKey` directly.

**Comment by HowardHinnant** [comment]
👍

**Comment by nbougalis** [comment]
Merged as 39f91351046bdff30b153f8442b562a3abe0ac82

**Pull request closed without merging**